src/middleware/status-handler.ts

```diff
@@ -0,0 +1,163 @@
+            return {
+                response: data,
+                succeeded: true,
+            };
+        })
+        .catch((e): CheckAccessibleRequestResponse => {
+            logger.error(`Failed to get ${name} status`, e);
+            let result;
+            if (isAxiosError(e)) {
+                result = e.response?.data?.data ?? e.response?.data;
+            }
+            result = result || e?.message || 'INACCESSIBLE';
+            return {
+                response: result,
+                succeeded: false,
+            };
+        })
+        .then((result: CheckAccessibleRequestResponse): DependencyObject => {
+            return {
+                ...result,
+                urlMD5: crypto.createHash('md5').update(url).digest('hex'),
+                name: name,
+            };
+        });
+};
+
+// We don't know the type at this point and we need to access it's members
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const hasPercentUp = (value: any): value is { percentUp: number } => (
+    typeof value === 'object' &&
+    'percentUp' in value &&
+    typeof value.percentUp === 'number'
+);
+
+interface StatusHandlerOptions {
+    versionPromise?: Promise<string | null>;
+    statusAccessibleOptions?: CheckAccessibleOptions[];
+    healthAccessibleOptions?: CheckAccessibleOptions[];
+    customChecks?: CustomCheckOptions[];
+}
+
+let alreadyCrawling = false;
+export const statusHandler = ({
+    versionPromise = Promise.resolve(null),
+    statusAccessibleOptions = [],
+    healthAccessibleOptions = [],
+    customChecks = []
+}: StatusHandlerOptions): Handler => async (req: Request, _res: Response, next: NextFunction): Promise<void> => {
+    const crawling = req.query.crawl === 'true';
+    try {
+        const version = await versionPromise;
+        const responseData: StatusResponse = {
+            version: version,
+            dependencies: [],
+            percentUp: 0
+        };
+
+        let statusPromises: Promise<DependencyObject>[] = [];
+        let healthPromises: Promise<DependencyObject>[] = [];
+        if (crawling) {
+            if (alreadyCrawling) {
+                next(Boom.tooManyRequests('Cannot crawl while already crawling'));
+                return;
+            }
+            alreadyCrawling = true;
+            statusPromises = statusAccessibleOptions.map(statusAccessibleOption => checkAccessible(statusAccessibleOption));
+            healthPromises = healthAccessibleOptions.map(healthAccessibleOption => checkAccessible(healthAccessibleOption));
+        }
+        const customPromises = customChecks.map(customCheck => customCheck.call());
+
+        const statusResults = await Promise.all(statusPromises);
+        const healthResults = await Promise.all(healthPromises);
+        const customResults = await Promise.all(customPromises);
+        const allResults = [...statusResults, ...healthResults, ...customResults];
```

Comment:
Should these be `allSettled`, or will checkAccessible always handle it?

Reply:
checkAccessible has a catch and thus cannot reject